src/main/webapp/app/entities/programming-exercise/actions/programmming-exercise-instructor-submission-state.component.ts

```diff
@@ -0,0 +1,121 @@
+import { Component, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
+import { catchError, debounceTime, map, tap } from 'rxjs/operators';
+import { ExerciseSubmissionState, ProgrammingSubmissionService, ProgrammingSubmissionState } from 'app/programming-submission/programming-submission.service';
+import { of, Subscription } from 'rxjs';
+import { NgbActiveModal, NgbModal } from '@ng-bootstrap/ng-bootstrap';
+
+/**
+ * This components provides two buttons to the instructor to interact with the students' submissions:
+ * - Trigger builds for all student participations
+ * - Trigger builds for failed student participations
+ *
+ * Also shows an info section next to the buttons about the number of building and failed submissions.
+ */
+@Component({
+    selector: 'jhi-programming-exercise-instructor-submission-state',
+    templateUrl: './programmming-exercise-instructor-submission-state.component.html',
+})
+export class ProgrammmingExerciseInstructorSubmissionStateComponent implements OnChanges, OnInit {
+    ProgrammingSubmissionState = ProgrammingSubmissionState;
+
+    @Input() exerciseId: number;
+
+    hasFailedSubmissions = false;
+    buildingSummary: { [submissionState: string]: number };
+    isBuildingFailedSubmissions = false;
+
+    resultEtaInMs: number;
+
+    submissionStateSubscription: Subscription;
+    resultEtaSubscription: Subscription;
+
+    constructor(private programmingSubmissionService: ProgrammingSubmissionService, private modalService: NgbModal) {}
+
+    ngOnInit(): void {
+        this.resultEtaSubscription = this.programmingSubmissionService.getResultEtaInMs().subscribe(resultEta => (this.resultEtaInMs = resultEta));
+    }
+
+    /**
+     * When the selected exercise changes, create a subscription to the complete submission state of the exercise.
+     *
+     * @param changes only relevant for change of exerciseId.
+     */
+    ngOnChanges(changes: SimpleChanges): void {
+        if (changes.exerciseId && !!changes.exerciseId.currentValue) {
+            this.submissionStateSubscription = this.programmingSubmissionService
+                .getSubmissionStateOfExercise(this.exerciseId)
+                .pipe(
+                    map(this.sumSubmissionStates),
+                    // If we would update the UI with every small change, it would seem very hectic. So we always take the latest value after 1 second.
+                    debounceTime(1000),
+                    tap((buildingSummary: { [submissionState: string]: number }) => {
+                        this.buildingSummary = buildingSummary;
+                        this.hasFailedSubmissions = this.buildingSummary[ProgrammingSubmissionState.HAS_FAILED_SUBMISSION] > 0;
+                    }),
+                )
+                .subscribe();
+        }
+    }
+
+    triggerBuildOfAllSubmissions() {
```

Comment:
Please add documentation

Reply:
You are right, also the naming of the method was bad.